sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+temporary USM memory but not carry around the pointer no longer have to
+synchronize with the execution of kernels accessing that memory and call the
+free function before continuing.
+
+A further benefit comes from the use of a memory pool for allocations, as
+malloc and free commands are asynchronous, the runtime can opportunistically
+re-use memory allocations from one kernel to another, without any intermediate
+freeing and re-allocation. This benefit can be further extended beyond queue
+synchronization by specify a threshold, which will instruct the runtime to (if
+possible) maintain a certain size of memory and not release that back to the OS
+even when the queue is synchronized with.
+
+There are also other properties which can be used when constructing a memory
+pool object to control other aspects of how the memory is allocated.
+
+== Integration with other extensions
+
+It is a priority for this extension to integrate with the
+link:../supported/sycl_ext_oneapi_graph.asciidoc[
+sycl_ext_oneapi_graph] extension. The asynchornous commands introduces in this
+extension are intended to be compatible with the SYCL Graph record and replay
+mode, however, the specifics of this will be defined in that extension.
+
+It is a priority for this extension to integrate with the
+link:../supported/sycl_ext_oneapi_bindless_images.asciidoc[
+sycl_ext_oneapi_bindless_images] extension. It is the intention that the memory
+allocations from the memory pool in this extension be used for the bindless
+images, however, this will need to be investigated further and so the details of
+this are not included in this version of the extension.
+
+== Examples
+
+There are various ways in which this extension can be used but a typical usage
+of the memory pool and the asynchronous malloc and free commands is described
+below. In this example an explicit memory pool is created and this is used to
+share memory allocated from the OS between multiple asynchornous malloc
+commands.
+
+[source,c++]
+----
+int main(int argc, char *argv[])
+{
+  queue q(property::queue::in_order);
+
+  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
+    q.get_device(), usm::alloc::device);
+  
+  {
+    // memory pool allocates memory from the OS
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for first kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  {
+    // memory pool re-uses previously allocated memory
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for second kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something_else(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // memory pool releases memory back to the OS
+  q.wait();
```

Comment:
the memory pool is created with q.context and q.device, it is not bound with queue itself. What will happen if multiple queues are used as below code.

```
queue q1(inorder), q2(inorder);
memory_pool memPool(q1.get_context(), q1.get_device(), usm::alloc::device);

void *temp1 = async_malloc_from_pool(q1, 1024, memPool);
async_free(q1, temp1);

void *temp2 = async_malloc_from_pool(q2, 1024, memPool);
async_free(q2, temp2);

q1.wait();
```

Reply:
This is a good question, this isn't specified in the extension right now. I would say that since the memory pool is tied to a context/device pair, `async_malloc`/`async_malloc_from_pool` should require that the device associated with the queue is one of the devices associated with the memory pool.

Reply:
I've clarified the wording of `async_malloc` and `async_malloc_from_pool` to state that the queue and the memory pool must be associated with the same context and device.